ark/phenotyping/som_utils.py

```diff
@@ -447,7 +623,246 @@ def visualize_cluster_data(fovs, channels, base_dir, cluster_dir, cluster_col='c
+        lr_start (float):
+            The start learning rate for the SOM, decays to lr_end
+        lr_end (float):
+            The end learning rate for the SOM, decays from lr_start
+        num_passes (int):
+            The number of training passes to make through the dataset
+        seed (int):
+            The random seed to set for training
+    """
+
+    # define the data paths
+    cell_table_path = os.path.join(base_dir, cell_table_name)
+    cluster_counts_path = os.path.join(base_dir, cluster_counts_name)
+    weights_path = os.path.join(base_dir, weights_name)
+
+    if not os.path.exists(cell_table_path):
+        raise FileNotFoundError('Cell table %s does not exist in base_dir %s' %
+                                (cell_table_name, base_dir))
+
+    # generate a matrix with each fov/cell label pair with their pixel SOM/meta cluster counts
+    print("Counting the number of pixel SOM/meta cluster counts for each fov/cell pair")
+    cluster_counts = compute_cell_cluster_counts(
+        fovs, base_dir, pixel_consensus_dir, cell_table_path, cluster_col
+    )
+
+    # write the created matrix
+    feather.write_dataframe(cluster_counts,
+                            os.path.join(base_dir, cluster_counts_name),
+                            compression='uncompressed')
+
+    # run the SOM training process
+    process_args = ['Rscript', '/create_cell_som.R', ','.join(fovs), str(xdim), str(ydim),
+                    str(lr_start), str(lr_end), str(num_passes), cluster_counts_path,
+                    weights_path, str(seed)]
+
+    process = subprocess.Popen(process_args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+    # continuously poll the process for output/error to display in Jupyter notebook
+    while True:
+        # convert from byte string
+        output = process.stdout.readline().decode('utf-8')
+
+        # if the output is nothing and the process is done, break
+        if process.poll() is not None:
+            break
+        if output:
+            print(output.strip())
+
+
+def cluster_cells(base_dir, cluster_counts_name='cluster_counts.feather',
+                  weights_name='cell_weights.feather', cell_cluster_name='cell_clusters.feather'):
+    """Uses trained weights to assign cluster labels on full cell data
+
+    Saves data with cluster labels to cell_cluster_name
+
+    Args:
+        base_dir (str):
+            The path to the data directory
+        cluster_counts_name (str):
+            Name of the file with the cluster counts of each cell
+            Created by train_cell_som
+        weights_name (str):
+            The name of the weights file, created by train_cell_som
+        cell_cluster_name (str):
+            The name of the file to write the clustered data
+    """
+
+    # define the paths to the data
+    cluster_counts_path = os.path.join(base_dir, cluster_counts_name)
+    weights_path = os.path.join(base_dir, weights_name)
```

Comment:
The `verify_in_list` call that's in `cluster_pixels` isn't here

Reply:
For cell clustering we don't need to `verify_in_list` the weights columns because unlike pixel clustering, the user is never going to subset on the pixel/meta cluster columns of `cluster_counts_path`.

Reply:
That's true, but if they used SOM clusters, instead of the metaclusters, or a different number of metaclusters, the columns will look different.